# Milestone II Project Report
## Cory Bilyeu and Sandro Bruno

### Motivation
Most diseases disrupt multiple proteins, and drugs treat such diseases by restoring the functions of the disrupted proteins. How drugs restore these functions, however, is often unknown because a drug’s therapeutic effects are not limited to the proteins that the drug directly targets. Thereby, we noticed two key observations that inspired our unsupervised Milestone II project drafted below. 

![image](https://user-images.githubusercontent.com/52717506/151834885-1b1a4f5f-5214-4064-abc7-e4e38ff8affc.png)

Therefore, we hypothesize that physical interactions between proteins alone cannot explain treatment since many drugs treat diseases by affecting the biological functions disrupted by the disease rather than directly targeting disease proteins or their regulators.

For this project we plan to develop a network that may enable us to explain proximal but especially (the less trivial) distal disease treatment effects. To do so we are planning to integrate physical protein-protein interactions, disease associated proteins, drug targets with and without biological functions into one network. In this network, each node represents one of the 6 entities: protein, disease-related protein (=disease causing proteins), drug-target protein (=proteins that are targeted by drugs), biological function (=ontologies about biological functions), disease or drug.

### Data Sources
The data for this network was downloaded using ‘wget’ functionality from the following places: disease-protein interactions (https://www.disgenet.org/; 3x tsv.gz files), protein-protein interactions (https://string-db.org/; 2x txt.gz ) (https://www.opentargets.org/; multiple json files), drug-protein interactions (https://www.ebi.ac.uk/chembl/; multiple json files) and biological functions (http://geneontology.org/; 1x .obo file ). All the 4 layers need to be mapped to the represented unique identifiers in each of the layers. We used the gene.info API (https://mygene.info/) to convert the different gene identifiers of each layer to map to the protein layer accordingly as well as the biological functions. For the conversion between durg approved indication and the disease layer we used a lookup table from DisGeNet (https://www.disgenet.org/; 1x tsv.gz).

### Network Creation and Data Manipulation
All the network subgraph visualizations were done in NetworkX, but the full network was visualized using gephi 0.92 due to the size of the network and computational power required to do so. The following is the largest connected component of the full network we created without any filtering or biological functions, and it consists of 28,208 nodes and 129,976 edges of various protein-protein interactions, drug-protein interactions, and disease protein interactions.

![image](https://user-images.githubusercontent.com/52717506/151835030-02c04247-d6be-424c-ab12-67772d5d1002.png)

### Physical Protein-Protein Interaction Layer
We have generated a network consisting of 132,518 (edges) physical protein-protein interactions between 11,263 (nodes) unique proteins by leveraging the STRING database that extracts its experimental data from 7 independent resources (BIND, DIP, GRID, HPRD, IntAct, MINT, and PID). We only considered human proteins and their interactions; only allowed protein–protein interactions with direct experimental evidence which are experimentally proven with a confidence score above 0.4 (commonly used threshold in the community; score ranges from 0-1) to include most reliable interactions; therefore filtering out genetic and indirect interactions between proteins such as those identified via synthetic lethality experiments. Last but not least we have mapped the Ensembl protein IDs to Entrez IDs using the gene info API in order to harmonize the nodes with the other network layers.   

### Protein-Disease Interaction Layer
We have used DisGeNet, one of the largest publicly available collections of genes and variants associated with human diseases to map diseases to genes they affect through effects like genomic alterations, altered expression, or post-translational modification. In order to ensure most reliable disease-gene associations, we only consider the curated set of disease-gene associations provided by DisGeNet which draws from expert-curated repositories such as UNIPROT, CGI, ClinGen, Genomics England, CTD (human subset), PsyGeNET, and Orphanet. Thereby, we excluded all disease-gene associations that are inferred, based on orthology relationships from animal models or based on computational-mining of the literature in order to denoise the network as we consider these as less confident edges. 

To avoid circularity in the random walk analysis, we removed disease-gene associations marked as therapeutic (Otherwise, the random walk would be more likely to terminate already after one step since that protein links to a drug that serves as a terminal node when starting from the disease and vise-versa). Furthermore, We removed indication nodes which are redundant or too unspecific by removing diseases with “Diseases” in their name such as “Autoimmune Diseases”, “Virus Diseases”, “Metabolic Diseases” or removing diseases with more than 60 approved drugs (i.e., Neoplasms with 245 or Cardiovascular Diseases with 184 approved drugs). The threshold of 60 was used because it seemed after iterative investigation that it removed most of the generic and non specific diseases. In contrast below that threshold we identified diseases which were more specific and therefore less favourable.

### Protein-Drug Interaction Layer
To ensure universal and unique mapping for the nodes in the protein-drug interaction layer, we utilized the mygene api to convert drug target Ensembl protein IDs to Entrez gene IDs. The method was utilized for all the drug target nodes, but there are no entrez gene ids for the drug nodes, so instead we maintained the ChEMBL ids for each molecule to use as unique identifiers for drug nodes within the network. There were many different action types for the drugs within the network, but we decided to filter the drugs to only those which were inhibiting the target protein (INHIBITOR, ANTAGONIST, BLOCKER, NEGATIVE ALLOSTERIC MODULATOR, INVERSE AGONIST, NEGATIVE MODULATOR, DEGRADER, ANTISENSE INHIBITOR). We considered this step to be very critical for the interpretation of our results since the network is undirected. This means we would not be able to know if a drug is inducing or preventing a disease by just looking at the similarity of two diffusion profiles which are similar to each other. By considering only inhibitors it will be easier for us to judge the directionality when the two embeddings are very similar to each other.

### Biological Function–Biological Function Interactions
We have downloaded the basic version of the Gene Ontology knowledge base, the world’s largest source of information on the functions of genes. The retrieved network represents a curated hierarchy of biological functions, where highly specific biological functions are children of more general biological functions according to numerous relationship types. We filtered such that the graph is guaranteed to be acyclic and annotations can be propagated up the graph. We included only relations between the biological functions that are specifically related to treatments such as is_a, part_of, regulates, negatively_regulates and positively_regulates. For example, “negative regulation of response to interferon-gamma” is_a  “negative regulation of innate immune response” is_a “negative regulation of immune response” negatively regulates  “immune response”. In addition, we only consider biological functions which are associated with at least one drug target or one disease protein, either directly or implicitly through their children in order to allow the model to focus on the biological functions most relevant to treatments.

### Protein–Biological Function Interactions
We map biological functions to the proteins they affect by using the gene.info API (12,488 proteins; 12,771 biological functions; 57,386 edges). Thereby, we only allowed experimentally verified associations between proteins and biological functions according to the following IDs: EXP (=inferred from experiment), IDA (=inferred from direct assay), IMP (=inferred from mutant phenotype), IGI (=inferred from genetic interaction), HTP (=high throughput experiment), HDA (=high throughput direct assay), HMP (=high throughput mutant phenotype), and HGI (=high throughput genetic interaction) and excluded any protein–biological function relationships that are inferred from physical interactions to avoid redundancy with the physical protein network of interacting proteins. We also excluded interactions inferred from gene expression patterns since the Gene Ontology states that such interactions are challenging to map to specific proteins. To prevent circularity, we further ignore all associations based on phylogenetically inferred annotations or various computational analyses (sequence or structural similarity, sequence orthology, sequence alignment, sequence modeling, genomic context, reviewed computational analysis) as we consider these as less confident links. Finally, we ignore associations based on author statements, curator inference, electronic annotations (i.e., automated annotations), and those for which no biological data was available. Last but not least, we filtered all interactions that were isolated in the network and did not connect to the main network. 

### Random Walk
To identify how a drug treats a disease, our approach uses an unbiased random walk algorithm (csrgraph memory efficient random walk implementation) that we fine tune such that it only starts at a drug or disease node. At every step, the walker can restart its walk or jump to an adjacent node. Whenever the random walker jumps to a starting node (drug or disease) we restart the random walk because drugs and diseases do not propagate their effect by using other drugs and diseases as intermediates. We then count how often each node was visited in the network, thus embedding the effect of the drug or disease on every protein and biological function, known as diffusion profile.
Such diffusion profiles provide at least two methodological advances. First, diffusion profiles provide a flexible and general framework to adaptively integrate physical interactions between proteins and a hierarchy of biological functions. Second, diffusion profiles provide a mathematical formalization of how drug and disease effects might propagate in a biological network. Thereby, we consider two possibilities how the effect might be propagated. First, proteins and biological functions closer to the drug or disease are affected more strongly. Similarly in diffusion profiles, proteins and biological functions closer to the drug or disease are visited more often since the random walker is more likely to visit them after a restart. Second, the net effect of the drug or disease on any given node depends on the net effect on each neighbor. Similarly in diffusion profiles, a random walker can arrive at a given node from any neighbor. 

### Cosine Similarity
Therefore, for a drug to treat a disease, it must affect proteins and biological functions similar to those disrupted by the disease. The diffusion profiles of the drug r(dr) and the disease r(di) encode the effect of the drug and the disease on proteins and biological functions. Therefore, comparing r(dr) and r(di) in theory should allow us to assess what drugs treat a given disease. For the comparison between two different diffusion profiles we used the cosine similarity because even if two similar diffusion profiles are far apart by the Euclidean distance e.g because of the pur frequency scale of visited nodes (e.g. one profile contain much more jumps/steps then the second walker since the terminations due to dense terminal disease or drug nodes were higher for the second profile) they could still have a smaller angle between them which is independent of the scale/frequency of the visits (therefore kind of normalize the random walks to each other - the smaller the angle the higher the similarity - we consider this problem to be very similar to a comparison between two documents which you do not want to be highly dependent on the number of words in a document but rather on the semantic directional similarity for example).

### Clustering and Visualization
Next, we clustered the cosine similarity matrix between any possible drug and disease pair using different approaches ranging from Hierarchical clustering via K-means to density based spatial clustering (DBSCAN) methodologies with and without dimensional reduction methods such as PCA, UMAP, random projections or tSNE.
In order to assess the quality of the resulting clusters we determined if the clusters are enriched with the clinically approved drug indication pairs or certain indication classes. Thereby, we applied the 2-way ANOVA (similarity~approval_status+clusters+clusters:approval_status) to determine the main effects and followed that assessment with the Tukey Post Hoc test to judge the over or underrepresentation of clinically approved drug indication pairs or Indication classes in each specific cluster (Note we have never encoded the drug approved indication pairs). Last but not least, we have reviewed the cases where we have the strongest and worst overlap between drug and indication pairs.     

### Results
We have generated two comprehensive networks that consist of 4 layers, plain experimentally verified physical protein-protein interactions, disease or drugs as well as biological functions. Overall, our two networks provide a large, systematic dataset to study drug–disease treatments. Nearly 1,000 approved treatments (i.e. drug–disease pairs) spanning almost every category of human anatomy are compiled, exceeding the largest prior network-based study by 4x (Guney et al 2016). 

To learn how the effects of drugs and diseases propagate through proteins and biological functions, we harnessed network diffusion profiles. Such a network diffusion profile was calculated by customizing an unbiased random walk. 

By comparing (Fig 3, using different methods PCA, HC …) the cosine similarity of drug and disease diffusion profiles we have noticed that there are strong intra similarities opposed to inter similarities that seemed to mask the effects we were looking for (drug to disease or disease to drug similarity). In other words drugs are more similar to other drugs than diseases or diseases are more similar to other diseases than drugs and therefore cluster better to each other. One part of a potential explanation is that there are multiple drugs that target the same drug target (protein) and therefore have almost identical embeddings since they have the same starting point for the random walk (Fig 3, indicated via strong diagonal purple clusters of intra similarities). Therefore, we have removed all drug to drug or disease to disease pairs which allowed us to better evaluate the network quality in a biological sense.
![image](https://user-images.githubusercontent.com/52717506/151835539-8da68d6f-2124-4961-bf51-e69a0d887936.png)
Do we see an enrichment of drug approved indication pairs opposed to non approved drug to indication pairs? 

This we have addressed by determining whether the drug approved indication pairs have significantly higher similarities than non-approved drug indication pairs. Otherwise, there is no sense to further explore our networks to explore the propagated drug treatment effects. As seen in Fig 4, we could observe in both networks significantly higher similarities in drug-approved indications opposed to non-approved indication pairs. However, we need to mention that the specific similarities are pretty low in general (~ 0.15-0.2 median; note 0 = independence). We believe that this is mainly because we relate highly dimensional and sparse vectors to each other and did not use weights to guide the random walk. However, in such a short project it was just not feasible to adjust the weights for such a big network - therefore, we have continued to work with what we have, an undirected/unweighted network. 

![image](https://user-images.githubusercontent.com/52717506/151835640-8f5c3b20-29cc-41d4-9d1c-706a35cfd45a.png)

Next, we have clustered the symmetric cosine similarity matrix and compared different techniques with and without the reduction of dimension (note: the embeddings are of high dimensionality and sparsity). The best performing clustering process was to reduce the dimensionality of the embeddings using UMAP followed by a Density-based spatial clustering. We are aware that this process is very controversial in the data science community since UMAP is only partially preserving the density. However, this kind of technique is well accepted in the bioinformatics field and also widely applied to biological data such as scRNAseq. Second there is also evidence in the data science field that in some scenarios it outperforms other preprocessing steps such as PCA to denoise the data (similar than for our project here see here).
![image](https://user-images.githubusercontent.com/52717506/151835743-973cc99c-624f-4d4e-ade2-56e6ed0aa4e3.png)

After we have formed the clusters we also have determined if certain clusters were enriched for drug approved indication pairs using a 2 way anova followed by a post hoc Tukey test pointing to clusters which were more enriched and others that were less enriched. If we would have more time, we would have definitely looked into both classes of clusters to see why they might be so different. We have also explored if there was an enrichment in certain indication classes (e.g. skin, neuron disease …) for the different clusters (data not shown). However, there was no such enrichment/bias observed.
![image](https://user-images.githubusercontent.com/52717506/151835793-34e77ddb-0463-45fd-a6c3-d3ac49e81d20.png)

Finally, we wanted to check if our hypothesis would hold. Can we improve the similarity between diffusion profiles of drug treatments and diseases by adding biological functions to bridge between distal drug targets and disease related proteins, therefore reducing the distance between the two profiles? 

In order to address this question we determined the shortest paths for drug approved indication pairs in both networks and subtracted these from each other (Fig 7; on x-axis). Next, we determined the rank for each approved drug-indication pair by its ascending order (lowest similarity between approved drug disease pair has rank 1 while the highest similarity between drug and disease has rank N, where N is the number of approved drug-disease pairs) and subtracted the two ranks from each of the networks from each other represented on y axis of Fig 7. As you can see, there is no advantage in adding the biological functions on our network to reduce the distance and improve the similarity under the aforementioned  parameters/assumptions (= most pairs are at 0; therefore shortest path did not decrease using the biological functions). However, based on the motivational observations for this project described in Fig 1 we would have expected that there are in an ideal world at least 28% (right side of Fig 1; 0.4*0.7=0.28) of drug approved indication pairs that could benefit from bridging biological functions. However, we found that only in one case indicated by a reduction of the shortest paths and a rank improvement by ~360 (~+733% similarity increase). We speculate that the general lack of improvement is mainly due to the fact that we did not implement any weights to guide the random walk to jump from one layer to another (protein-protein - biological functions or vise versa).

![image](https://user-images.githubusercontent.com/52717506/151835903-e7a5afd5-17ba-4aaf-bcac-7767c69c341b.png)

Next we have looked into the one case that worked as we have envisioned (shown in Fig 8). As you can see it is perfectly in line how we would have liked to see it for all the other distal drug indication pairs. The biological function, “neurotransmitter catabolic process” replaced and summarized the two proteins “VDAC1” and “CANX” that are involved in the metabolic intervention of the ”Valpronic acid” drug treating ”Manic Disorder”. We are aware that this is just one example and maybe just by chance. However, it is encouraging to see that it could work in principle. In the future we would definitely plan to use weights to guide the random walk. Thereby, one could train the weights based on the efficacy data of the drug efficacy data of the clinical phase III studies for each approved drug indication pair. We have already downloaded the data - but did not have time to explore it further.  

![image](https://user-images.githubusercontent.com/52717506/151835968-d45d35b6-f00f-4665-b358-4d29ae4e439f.png)


